src/Toolkit/Toolkit/FileDownloadTask.cs

```diff
@@ -0,0 +1,804 @@
+        {
+            if (!IsResumable)
+                return RestartAsync();
+            Status = FileDownloadStatus.Resuming;
+            FileInfo f = new FileInfo(TempFile);
+            long offset = f.Exists ? f.Length : 0;
+            return BeginDownload(offset);
+        }
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Aborts the running download and deletes the partial download.
+    /// </summary>
+    /// <returns></returns>
+    public Task CancelAsync()
+    {
+        var task = transferTask;
+        if (Status == FileDownloadStatus.Paused)
+        {
+            if (File.Exists(TempFile))
+                File.Delete(TempFile);
+            BytesDownloaded = 0;
+            Status = FileDownloadStatus.Cancelled;
+            return Task.CompletedTask;
+        }
+        else
+        {
+            lock(_queueLock)
+            {
+                if (_inProgressItems.Contains(this))
+                {
+                    _inProgressItems.Remove(this);
+                }
+                if (_queuedItems.Contains(this))
+                {
+                    _queuedItems.Remove(this);
+                }
+            }
+            if (cancellationSource is null)
+            {
+                Status = FileDownloadStatus.Cancelled;
+            }
+            else
+            {
+                cancellationSource?.Cancel();
+            }
+            if (task is not null)
+            {
+                return task.ContinueWith(t => { File.Delete(TempFile); BytesDownloaded = 0; Status = FileDownloadStatus.Cancelled; ProcessQueue(); });
+            }
+            else
+            {
+                if (task is null && File.Exists(TempFile))
+                    File.Delete(TempFile);
+                ProcessQueue();
+                return Task.CompletedTask;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Pauses the active download.
+    /// </summary>
+    /// <returns></returns>
+    public Task PauseAsync()
+    {
+        lock (_queueLock)
+        {
+            if (_queuedItems.Contains(this))
+                _queuedItems.Remove(this);
+        }
+
+        var task = transferTask;
+        if (cancellationSource == null || task == null)
+        {
+            Status = FileDownloadStatus.Paused;
+            return Task.CompletedTask;
+        }
+        cancellationSource.Cancel();
+        return task.ContinueWith(t => { Status = FileDownloadStatus.Paused; ProcessQueue(); });
+    }
+
+    /// <summary>
+    /// Returns when the download completed (or failed)
+    /// </summary>
+    /// <returns></returns>
+    public async Task DownloadAsync()
+    {
+        if (Status == FileDownloadStatus.Paused || Status == FileDownloadStatus.Error)
+            await ResumeAsync().ConfigureAwait(false);
+        else if (transferTask == null)
+        {
+            throw new InvalidOperationException();
+        }
```

Comment:
It seems this will throw if FileDownloadTask is queued -- is that right?  Awaiting a queued-up file to finish downloading seems like a reasonable scenario.

Also, what about tasks in "Cancelled" status -- should they automatically call ResumeAsync (the way "Error" tasks do) or fail with an exception?